### ISSUE TITLE
Money is sent/requested multiple times if the enter key is pressed multiple times reported by @adeel0202

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#12714]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Actions Performed:

1. Go to any chat
2. Click on the plus icon
3. Click Send money or Request Money
4. Enter any amount
5. Click Next
6. Press enter key multiple times

## Expected Result:
Money is sent only once

## Actual Result:
Money is sent multiple times

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- Desktop App
- Mobile Web
- iOS
- Android (should be reproducible, right now it's crashing in the latest build which is a kI)

**Version Number:** 1.2.27-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/201730915-2bfd3b1f-8e65-4290-bf28-0151ad0a578a.mov
https://user-images.githubusercontent.com/43996225/201730938-ef44e1a4-8ef5-42f9-95ee-f9c3ae4383a7.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @adeel0202 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1668413790082549

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Puneet-here** [comment]
### Proposal

The issue is also reproducible while splitting bill or requesting money

To fix this we can show a loader at button like this
```diff
                    ? (
                        <SettlementButton
                            isDisabled={shouldDisableButton}
                            onPress={this.confirm}
                            shouldShowPaypal={Boolean(recipient.payPalMeAddress)}
                            enablePaymentsRoute={ROUTES.IOU_SEND_ENABLE_PAYMENTS}
                            addBankAccountRoute={ROUTES.IOU_SEND_ADD_BANK_ACCOUNT}
                            addDebitCardRoute={ROUTES.IOU_SEND_ADD_DEBIT_CARD}
                            currency={this.props.iou.selectedCurrencyCode}
+                           isLoading={this.props.iou.loading}
                        />
                    ) : (
                        <ButtonWithMenu
+                           isLoading={this.props.iou.loading}
                            isDisabled={shouldDisableButton}
                            onPress={(_event, value) => this.confirm(value)}
                            options={this.splitOrRequestOptions}
                        />
                    )}
```
https://github.com/Expensify/App/blob/c55d5e44eed2fb6216209458c1572cb8719f3740/src/components/IOUConfirmationList.js#L301-L312

**Comment by miljakljajic** [comment]
Adding external label

**Comment by miljakljajic** [comment]
@Beamanator , can you confirm this can be worked on by an external engineer?

@aimane-chnaif - take a look at Puneet's proposal when you have a chance. Thank you!

**Comment by Emz27** [comment]
### Proposal
- add debounce/throttle to onPress event using lodash

### Cause
- because of render lag, the javascript code cant keep up with native events

@Puneet-here proposal will help but It can still be replicated if the QA use their magic fingers since react state render is asynchronous

This can be fully solved with debounce/throttle on client + loader + backend limit

**Comment by Beamanator** [comment]
@miljakljajic Yep I agree this should be able to be worked externally 👍

**Comment by Emz27** [comment]
@aimane-chnaif heres the diff's

- we create debounce/throttle func instance upon creation of the page
- I chose throttle here because the debounce spawns a timer while throttle will just use timestamp
- the 800ms throttle time will give the component enough time to render loading and disabled button property changes which prevents any succeeding press

[throttle reference](https://underscorejs.org/docs/modules/throttle.html)

```diff
diff --git a/src/components/IOUConfirmationList.js b/src/components/IOUConfirmationList.js
index 4f4eb415e..d7f81de40 100755
--- a/src/components/IOUConfirmationList.js
+++ b/src/components/IOUConfirmationList.js
@@ -118,6 +118,7 @@ class IOUConfirmationList extends Component {
 
         this.toggleOption = this.toggleOption.bind(this);
         this.confirm = this.confirm.bind(this);
+        this.confirmThrottle = _.throttle(this.confirm, 800, {leading: true, trailing: false});
     }
 
     /**
@@ -273,7 +274,7 @@ class IOUConfirmationList extends Component {
     render() {
         const selectedParticipants = this.getSelectedParticipants();
         const shouldShowSettlementButton = this.props.iouType === CONST.IOU.IOU_TYPE.SEND;
-        const shouldDisableButton = selectedParticipants.length === 0;
+        const shouldDisableButton = selectedParticipants.length === 0 || this.props.iou.loading;
         const recipient = this.state.participants[0];
         const canModifyParticipants = !this.props.isIOUAttachedToExistingChatReport && this.props.hasMultipleParticipants;
 
@@ -282,7 +283,7 @@ class IOUConfirmationList extends Component {
                 sections={this.getSections()}
                 value={this.props.comment}
                 onSelectRow={canModifyParticipants ? this.toggleOption : undefined}
-                onConfirmSelection={this.confirm}
+                onConfirmSelection={this.confirmThrottle}
                 onChangeText={this.props.onUpdateComment}
                 textInputLabel={this.props.translate('iOUConfirmationList.whatsItFor')}
                 placeholderText={this.props.translate('common.optional')}
@@ -300,18 +301,20 @@ class IOUConfirmationList extends Component {
                     ? (
                         <SettlementButton
                             isDisabled={shouldDisableButton}
-                            onPress={this.confirm}
+                            onPress={this.confirmThrottle}
                             shouldShowPaypal={Boolean(recipient.payPalMeAddress)}
                             enablePaymentsRoute={ROUTES.IOU_SEND_ENABLE_PAYMENTS}
                             addBankAccountRoute={ROUTES.IOU_SEND_ADD_BANK_ACCOUNT}
                             addDebitCardRoute={ROUTES.IOU_SEND_ADD_DEBIT_CARD}
                             currency={this.props.iou.selectedCurrencyCode}
+                            isLoading={this.props.iou.loading}
                         />
                     ) : (
                         <ButtonWithMenu
                             isDisabled={shouldDisableButton}
-                            onPress={(_event, value) => this.confirm(value)}
+                            onPress={(_event, value) => this.confirmThrottle(value)}
                             options={this.splitOrRequestOptions}
+                            isLoading={this.props.iou.loading}
                         />
                     )}
             />
```

**Comment by bondydaa** [comment]
why does our API and Auth allow this? This should be fixed at every level of the stack not just in JS.

**Comment by bondydaa** [comment]
going to take this over and investigate it a bit more since I suspect we need to update the API and Auth.

**Comment by bondydaa** [comment]
sorry haven't had a chance to dig into this yet, been focused on the Upwork Integration auth command and had a bunch of ring1/IFR chores last week. Will get started on this tomorrow at the latest

**Comment by bondydaa** [comment]
oooh I see where my misunderstanding was... this is manually sending money to another user not paying an existing IOU (I though it was the latter).

While digging in here it looks like these commands have yet to be updated to use the Optimistic patterns for Offline usagehttps://github.com/Expensify/App/blob/3cf717412cf6b03357184bf057921cf658ff158d/src/libs/actions/IOU.js#L687-L689

so I went digging to see what was up with that and I see we have this issue for tracking those refactors https://github.com/Expensify/Expensify/issues/213712

and this PR from @youssef-lr https://github.com/Expensify/App/pull/12739/ which is going to drastically change how this functions. 

So I think we should just include a fix in that PR or at the very least wait until that is merged and re-test this. 

Ultimately I don't agree that we should throttle/debounce here the press/click events: that just delays the possibility of this happening it doesn't prevent it. 

I think the problem is:
It seems the button that submits the request and calls the API still has focus or is able to be triggered for whatever reason so that's why clicking enter triggers new requests, since the modal has been removed from the page we shouldn't allow it to be triggered again.

**Comment by Emz27** [comment]
@bondydaa Hi, my proposal solves the problem you mentioned. I used throttle func to limit the action once per 800ms, not delay it. It takes the first action then ignore the rest for 800ms. I used a mouse clicker macro to test this. I agree that this needs to be handled in api but an additional protection in client would be nice too.

**Comment by JmillsExpensify** [comment]
In the related PR from @youssef-lr, this issue is no longer reproducible. Based on that, we should either put this issue on hold and double confirm when the related PR is merged, or we should close.

**Comment by miljakljajic** [comment]
Closing!

**Comment by JmillsExpensify** [comment]
I agree with that.

**Comment by bondydaa** [comment]
threw a hold on here and dropped it to weekly but honestly this should just be closed - the PR is completely changing how this functions and doesn't wait for a response since it's made optimistically so the UI should have nothing to wait on and therefore no opportunity for the same request to be made multiple times.

Even if something like this is still present we will not implement a debounce to fix it, deboucing should only be used with user input like typing/entering text that would automatically submit a form. Here's what Tim has laid out as reasons _for_ a debounce (internal Expensify Slack link sorry https://expensify.slack.com/archives/C03TQ48KC/p1668007475845529?thread_ts=1667971989.964359&cid=C03TQ48KC)
<img width="385" alt="image" src="https://user-images.githubusercontent.com/4073354/204857125-d970c2fb-05be-4b25-a95a-ee9f07badc61.png">

With the current proposal what happens if the request takes longer than 800 miliseconds? then this is still possible - so that is not a fix it's just pushing the bug off until later so the proper fix is using Optimistic requests which is what the PR we are holding on is doing.

**Comment by Emz27** [comment]
I disagree with the points here, this will just establish a thinking of limited use for throttle/debounce and just ignore the asynchronous nature of react state and navigation. No matter how hard you optimize it, there will always still be lag and you cant expect the user's device to be fast. What I think is the reason enough to merit the use of throttle/debounce is if it is doing an expensive action, only once per process and it was reported as a bug for multiple trigger.

It doesnt matter if the request took longer than 800ms, it will still work. The 800ms here is just the allotment for the time the state changes from not loading to loading state which is more than enough. After that, the loading state will disable the button and will prevent any subsequent actions

**Comment by trjExpensify** [comment]
👋 That PR is on prod now, so @miljakljajic @bondydaa this can be re-tested. FWIW, I gave it a go on Web Chrome `v1.2.41-1` and can't reproduce it. 

![1BDfeOKC90](https://user-images.githubusercontent.com/16232057/208121214-897e1a3e-a403-4464-93e7-2f281ce01669.gif)

P.s - that's one my kittens enjoying seeing snow for the first time :cat:

**Comment by youssef-lr** [comment]
@aimane-chnaif can you try requesting money as well?

**Comment by youssef-lr** [comment]
Yeah I can reproduce on iOS as well, and in the request money flow as well.

**Comment by bondydaa** [comment]
it's unclear to me in those videos what is being pressed/tapped that is triggering the multiple requests.

**Comment by youssef-lr** [comment]
So the way the requests work right now in Onyx is that pressing the 'Settle elsewhere' button will trigger a call to `Aoi.write` which will push the request to Onyx queue. The processing that happens before pushing the request seems to take a bit longer in native and doesn't make the modal disappear immediately as in web. Probably due to the JS thread being busy (someone correct me if I'm wrong here).

Usually the solution to this if we were hitting the API right away would be to implement an `isLoading` condition, but it's not happening here and we'll need a similar solution to prevent subsequent requests from being pushed to the queue while the original one hasn't been processed yet.

The debounce solution would work here and if we were to implement I think it should be better done at Onyx level in `Api.write`, but I agree that it's not elegant and we should find a better way to handle this.

**Comment by bondydaa** [comment]
ah okay, thanks for clearing that up!

🤔 does this happen for other buttons around the product as well? if no, how are we handling them there? 

On web the request modal closes after clicking the button to submit pretty quickly but then if i try to click submit many times I can trigger this too. 

And I agree with @youssef-lr's assessment too. I think in most cases we tend to use a `isLoading` prop to disable a button or show a loader that would prevent a user repeating an action. 

I think we would probably prefer that solution over the debounce though I agree the debounce would probably work too. 

To go the passing of a prop route like `isLoading` (which maybe isn't the greatest name) we'd just set the button to disabled if `isLoading=true` which should give the modal time to close and prevent additional taps/clicks.

**Comment by youssef-lr** [comment]
This solution worked for me, but I'm not sure we would want to go with it. What do you think @bondydaa? Linting is also complaining I can't use `then` inside a view and logic should be moved to the action.

<img width="869" alt="Screenshot 2022-12-16 at 19 08 20" src="https://user-images.githubusercontent.com/9680864/208161725-520767f3-f9ae-4f86-9d67-40b3452ce195.png">
<img width="857" alt="Screenshot 2022-12-16 at 19 09 21" src="https://user-images.githubusercontent.com/9680864/208161734-f927f0ad-d9ad-4a1b-9ad5-1700e39f1e77.png">

**Comment by youssef-lr** [comment]
Forgot to include the change from [here](https://github.com/Expensify/App/blob/main/src/libs/actions/IOU.js#L214).
<img width="702" alt="Screenshot 2022-12-16 at 19 16 04" src="https://user-images.githubusercontent.com/9680864/208162974-05c7bc6f-3ede-4e64-b9b3-2f3a34dd45c6.png">

**Comment by Emz27** [comment]
Hi, is my proposal not valid anymore? I also used isLoading state but with throttle function. I still dont know why debounce sounds bad to you guys, if the debounce delays the action then it is bad, I agree with it but my proposal is using throttle function which doesnt delay any action, it just limits it. I still insist that the throttle may be necessary. I understand that it is not a fix and it just workaround but it works and the code changes is not that bad/big. I appreciate it if you guys can explain why it is bad. Also you would want to put a debounce/throttle action in component level  if you would go to that route because if you put it on api, that will restrict the api and you will be more likely to create another api for the same feature

**Comment by youssef-lr** [comment]
Hi @Emz27 - We haven't settled on any solution yet. If we can't come up with a better solution we will probably resort to using a debounce or throttle function.

**Comment by youssef-lr** [comment]
Here's an alternative solution, considering that the `IOUModal` component is mounted everytime we open the IOU modal, we can rely on the component's state to tell us if a transaction has been submitted, this should work for both requesting and sending money. Here's a diff:


```diff
diff --git a/src/pages/iou/IOUModal.js b/src/pages/iou/IOUModal.js
index 2f9bb14530..d3cad4f7f5 100755
--- a/src/pages/iou/IOUModal.js
+++ b/src/pages/iou/IOUModal.js
@@ -131,6 +131,7 @@ class IOUModal extends Component {
             // amount is currency in decimal format
             amount: '',
             comment: '',
+            isTransactionDone: false,
         };
 
         // Skip IOUParticipants step if participants are passed in
@@ -282,6 +283,9 @@ class IOUModal extends Component {
      * @param {String} paymentMethodType
      */
     sendMoney(paymentMethodType) {
+        if (this.state.isTransactionDone) {
+            return;
+        }
         const amount = Math.round(this.state.amount * 100);
         const currency = this.props.iou.selectedCurrencyCode;
         const comment = this.state.comment;
@@ -296,7 +300,6 @@ class IOUModal extends Component {
                 this.props.currentUserPersonalDetails.login,
                 participant,
             );
-            return;
         }
 
         if (paymentMethodType === CONST.IOU.PAYMENT_TYPE.PAYPAL_ME) {
@@ -309,12 +312,16 @@ class IOUModal extends Component {
                 participant,
             );
         }
+        this.setState({isTransactionDone: true});
     }
 
     /**
      * @param {Array} selectedParticipants
      */
     createTransaction(selectedParticipants) {
+        if (this.state.isTransactionDone) {
+            return;
+        }
         const reportID = lodashGet(this.props, 'route.params.reportID', '');
 
         // IOUs created from a group report will have a reportID param in the route.
@@ -349,6 +356,8 @@ class IOUModal extends Component {
             this.props.currentUserPersonalDetails.login,
             selectedParticipants[0],
             this.state.comment);
+
+        this.setState({isTransactionDone: true});
     }
 
     renderHeader() {

```

I'm curious what you guys think about this.

**Comment by youssef-lr** [comment]
I'm not sure about setting a property in the class itself...can you try this diff on the HT account? 
```diff
-                                                onConfirm={this.createTransaction}
-                                                onSendMoney={this.sendMoney}
+                                                onConfirm={(selectedParticipants) => {
+                                                    if (this.state.isTransactionDone) { return; }
+                                                    this.setState({isTransactionDone: true});
+                                                    this.createTransaction(selectedParticipants);
+                                                }}
+                                                onSendMoney={(paymentMethodType) => {
+                                                    if (this.state.isTransactionDone) { return; }
+                                                    this.setState({isTransactionDone: true});
+                                                    this.sendMoney(paymentMethodType);
+                                                }}

```

**Comment by youssef-lr** [comment]
> (this.setState solution proposed https://github.com/Expensify/App/issues/12714#issuecomment-1357108728 still doesn't work (when test with HT account or app becomes heavy or when tap with multiple fingers simultaneously) because of render lag explained https://github.com/Expensify/App/issues/12714#issuecomment-1315726590

I think it's because I'm setting that value to true only after requesting or sending money, if the call happens before as in this diff I don't think the issue will still occur:

```diff
-                                                onConfirm={this.createTransaction}
-                                                onSendMoney={this.sendMoney}
+                                                onConfirm={(selectedParticipants) => {
+                                                    if (this.state.isTransactionDone) { return; }
+                                                    this.setState({isTransactionDone: true});
+                                                    this.createTransaction(selectedParticipants);
+                                                }}
+                                                onSendMoney={(paymentMethodType) => {
+                                                    if (this.state.isTransactionDone) { return; }
+                                                    this.setState({isTransactionDone: true});
+                                                    this.sendMoney(paymentMethodType);
+                                                }}

```

But you bring up an interesting point, the use of state here maybe overkill because we don't render anything based on this condition. So we might be better off setting a property in the class itself.

**Comment by youssef-lr** [comment]
@0xmiroslav I think we have all agreed to go with your proposal. I would just prefer we use `creatingIOUTransaction` as it's more accurate, because we are not loading anything. I think it's also cleaner to have this logic as in the diff below to keep those methods clean:

```diff
-                                                onConfirm={this.createTransaction}
-                                                onSendMoney={this.sendMoney}
+                                                onConfirm={(selectedParticipants) => {
+                                                    if (this.creatingIOUTransaction) { return; }
+                                                    this.creatingIOUTransaction = true;
+                                                    this.createTransaction(selectedParticipants);
+                                                }}
+                                                onSendMoney={(paymentMethodType) => {
+                                                    if (this.creatingIOUTransaction) { return; }
+                                                    this.creatingIOUTransaction = true;
+                                                    this.sendMoney(paymentMethodType);
+                                                }}

```

What do you think?

**Comment by miljakljajic** [comment]
Apologies, @aimane-chnaif. I will make the relevant payments today.

**Comment by miljakljajic** [comment]
Contracts extended @adeel0202 @0xmiroslav @aimane-chnaif

**Comment by miljakljajic** [comment]
That's no problem, thank you for accepting the contract. I've paid you now and I'll end the contract.

**Comment by miljakljajic** [comment]
@aimane-chnaif once you accept the offer we can close this out. Thank you!

**Comment by miljakljajic** [comment]
Confirmed! Sent offer including timeline bonus.

**Comment by miljakljajic** [comment]
@aimane-chnaif if you accept the offer I can pay you and get this closed out!

**Comment by miljakljajic** [comment]
Paid and incorrect offers withdrawn, thank you @aimane-chnaif.

**Comment by miljakljajic** [comment]
Apol

**Comment by miljakljajic** [comment]
Paid, contract ended - closing out again.
